feat(bank)!: Move the bank module SendEnabled info into state (from Params).

dwedul-figure opened the pull request:
<!--
The default pull request template is for types feat, fix, or refactor.
For other templates, add one of the following parameters to the url:
- template=docs.md
- template=other.md
-->

## Description

Addresses most of: #11859 

1. Deprecates and renders unusable the x/bank module's `Params.SendEnabled` field.
2. Creates methods for getting and managing the `SendEnabled` information directly in the state store.
3. Creates a migration that moves the information from `Params.SendEnabled` into the state store.
4. Adds a `SendEnabled` field directly in the GenesisState struct.
5. Adds the ability to update a genesis file to the new version using `genutil`.
6. Allows reading the old genesis format into the new one.
7. Creates a GRPC query for getting the SendEnabled information.
8. Creates a new query command for getting this info on the command line.
9. ~~Creates a new `MsgSetSendEnabled` message for use with the governance module, allowing the setting of any number of denoms.~~ This has been split out to another PR: #11981 

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

alexanderbez commented on proto/cosmos/bank/v1beta1/query.proto:
```diff
@@ -63,6 +63,12 @@ service Query {
     option (google.api.http).get = "/cosmos/bank/v1beta1/denom_owners/{denom}";
   }
+
+  // SendEnabled queries for SendEnabled entries on a denomination.
+  // Denominations that don't have a specific SendEnabled entry, are not returned and would thus use the default (defined in Params).
```
Can we please keep the proto + godocs wrapped at 80 chars 🙏

alexanderbez replied:
> are not returned and would thus use the default (defined in Params).

What does this mean? The `Params.SendEnabled` field is deprecated and should not be used. What should be stated and implemented is that if a denom is not present, then the denom is allowed by default. Granted, we can also make the inverse assumption (i.e. NOT enabled by default).

alexanderbez replied:
Ahh yes, that makes sense.

alexanderbez commented on proto/cosmos/bank/v1beta1/query.proto:
```diff
@@ -241,3 +247,18 @@ message QueryDenomOwnersResponse {
+  // denoms is the specific denoms you want look up. Leave empty to get all entries.
+  repeated string denoms = 1;
+  // pagination defines an optional pagination for the request.
+  cosmos.base.query.v1beta1.PageRequest pagination = 99;
```
why 99?

alexanderbez commented on proto/cosmos/bank/v1beta1/query.proto:
```diff
@@ -241,3 +247,18 @@ message QueryDenomOwnersResponse {
+message QuerySendEnabledResponse {
+  repeated SendEnabled send_enabled = 1;
+  // pagination defines the pagination in the response.
+  cosmos.base.query.v1beta1.PageResponse pagination = 99;
+}
```
nit: new line

alexanderbez commented on proto/cosmos/bank/v1beta1/tx.proto:
```diff
@@ -16,6 +16,9 @@ service Msg {
   // MultiSend defines a method for sending coins from some accounts to other accounts.
   rpc MultiSend(MsgMultiSend) returns (MsgMultiSendResponse);
+
+  // SetSendEnabled is a governance operation for setting the SendEnabled flag on any number of Denoms.
+  rpc SetSendEnabled(MsgSetSendEnabled) returns (MsgSetSendEnabledResponse);
```
We shouldn't have this in this PR. Please make a seperate PR for governance operations to update these :)

dwedul-figure replied:
Is there a reason to split it out other than making a PR smaller? I ask because without this endpoint, there would be no way to set the `SendEnabled` state for denoms. Currently it's done through a governance proposal to set the bank module's Params. Having this is required in order to maintain current functionality.

dwedul-figure replied:
Sounds good. Give me a bit and I'll have it split out.

dwedul-figure replied:
Okay. I split it out the `MsgSetSendEnabled` stuff as PR #11981.

I also made a couple other small tweaks:
1. I default to using `nil` for the `Params.SendEnabled` slice (instead of an empty slice).
2. I updated the bank keeper's `SetParams` function to set the `SendEnabled` entries in the new way. That way, anyone using that method won't have weird things happen.

dwedul-figure commented on client/utils.go:
```diff
@@ -76,3 +77,24 @@ func ReadPageRequest(flagSet *pflag.FlagSet) (*query.PageRequest, error) {
+// ReadPageRequest expects it to be the raw bytes.
+//
+// Common usage: pageReq, err := client.ReadPageRequest(client.WithPageKeyDecoded(cmd.Flags()))
+func WithPageKeyDecoded(flagSet *pflag.FlagSet) *pflag.FlagSet {
```
I changed the name to `FlagSetWithPageKeyDecoded` and made it also return an error. I then made `MustFlagSetWithPageKeyDecoded` so that the 1-line option is still there.

dwedul-figure commented on client/utils.go:
```diff
@@ -76,3 +77,24 @@ func ReadPageRequest(flagSet *pflag.FlagSet) (*query.PageRequest, error) {
+func WithPageKeyDecoded(flagSet *pflag.FlagSet) *pflag.FlagSet {
+	encoded, err := flagSet.GetString(flags.FlagPageKey)
+	if err != nil {
+		panic(err.Error())
```
Change made. See https://github.com/cosmos/cosmos-sdk/pull/11977#discussion_r874940109

amaurym commented:
I propose to not include this change in v046, it's already big enough, and most parts are audited.

We should tag a beta3 soon, but block this until then.

dwedul-figure commented:
Okay. I split it out the `MsgSetSendEnabled` stuff as PR #11981.

I also made a couple other small tweaks:
1. I default to using `nil` for the `Params.SendEnabled` slice (instead of an empty slice).
2. I updated the bank keeper's `SetParams` function to set the `SendEnabled` entries in the new way. That way, anyone using that method won't have weird things happen.

dwedul-figure commented on proto/cosmos/bank/v1beta1/query.proto:
```diff
@@ -63,6 +63,15 @@ service Query {
+  // This query only returns denominations that have specific SendEnabled settings.
+  // Any denomination that does not have a specific setting will use the default
+  // params.default_send_enabled, and will not be returned by this query.
+  rpc SendEnabled(QuerySendEnabled) returns (QuerySendEnabledResponse) {
```
I renamed `QuerySendEnabled` to `QuerySendEnabledRequest` since the other messages in that proto also start with `Query`.

robert-zaremba commented on proto/cosmos/bank/v1beta1/bank.proto:
```diff
@@ -10,8 +10,12 @@ option go_package = "github.com/cosmos/cosmos-sdk/x/bank/types";
 message Params {
-  option (gogoproto.goproto_stringer)       = false;
-  repeated SendEnabled send_enabled         = 1;
+  option (gogoproto.goproto_stringer) = false;
```
It's almost correct. In the comment we need to add a version pointing from when it is deprecated.

dwedul-figure replied:
Updated the comment to
```proto
  // Deprecated: Use of SendEnabled in params is deprecated.
  // For genesis, use the newly added send_enabled field in the genesis object.
  // Storage, lookup, and manipulation of this information is now in the keeper.
  // As of v0.47.0, it only exists here for backwards compatibility of genesis files.
```

dwedul-figure replied:
For what it's worth, none of the other deprecated proto messages or fields have similar version statements.

It's tough to get it correct right now. There's just no way of knowing now, which version will contain this change.

Here's some option analysis:
1. Use "As of v0.47.0 ...".
   **PRO**: If it's released in v0.47.0, everything is good and we've provided a nice shortcut to this info.
   **CON**: If it's released later, it can cause confusion. E.g. A person sees the comment (in maybe v0.48.0), grabs v0.47.0 and it's not actually deprecated there.
2. Use "Only exists after v0.46 ...".
   **PRO**: No matter which version the change is released in, this would _technically_ be a correct statement.
   **CON**: If not released in v0.47.0, it can still cause the same confusion mentioned before. A person might see the statement in v0.48.0, then notice it missing from v0.47.0.
3. Don't set it now. Instead, create an issue to have the comment added closer to when it's released. That issue could also ask that similar comments be added to the other deprecated proto messages and fields.
   **PRO**: There's much less chance that it's incorrect.
   **CON**: Someone wondering which version it was deprecated in will have to take an extra step and dig through the change log.
4. Omit the version comment entirely.
   **PRO**: don't have to worry about getting it right.
   **CON**: anyone wanting to know the last usable version (or first unusable version) would have to search the change log.

I'm personally partial to picking point 3. My second choice would be 4. I'd rather not have the information there at all than to have it be incorrect. The info will be available in the change log either way, but contradictory info can cause problems.

One factor to weigh when comparing 1/2 vs 3/4 is the odds that this isn't released in v0.47.0. I feel like there's about a 10% chance of that. That's kind of based on the amount of stuff that's changing in v0.46.0, but mostly just gut feeling (i.e. completely made up). You all probably have more accurate knowledge and feelings on that.

alexanderbez replied:
I would go with `Only exists after v0.46 ...`, because I say 0.47 won't exist because we're actually aiming for 1.0

dwedul-figure replied:
Updated it to
```proto
// Deprecated: Use of SendEnabled in params is deprecated.
// For genesis, use the newly added send_enabled field in the genesis object.
// Storage, lookup, and manipulation of this information is now in the keeper.
// This version only still exists after v0.46 for backwards compatibility of
// genesis files.
```

alexanderbez replied:
We don't know what the next release will be, so we can't use 0.47, so that's why "next release after 0.46" makes the most sense.

alexanderbez commented on x/bank/keeper/genesis.go:
```diff
@@ -10,8 +10,14 @@ import (
 
 // InitGenesis initializes the bank module's state from a given genesis state.
 func (k BaseKeeper) InitGenesis(ctx sdk.Context, genState *types.GenesisState) {
+	genState.MigrateSendEnabled()
```
Why is this being called here? Migrations should happen explicitly in upgrade handlers.

dwedul-figure replied:
This allows continued use of a genesis file that has `SendEnabled` entries defined in the bank params.

I've updated genesis file creation to have any new genesis files created using the new field, but I'm unaware of any code location meant for migrating a genesis file. By having this line here, though, there's no need to migrate the file.

All `MigrateSendEnabled` does is move things around in the GenesisState object (from `GenesisState .Params.SendEnabled` to the new `GenesisState.SendEnabled` field). It doesn't update any files.

alexanderbez replied:
I see. I like how you thought of this, but this is not how it's supposed to work.

`InitGenesis` is supposed to work as if it's working on the latest valid state. This is why we have "migrations". You can see many examples of this in some of the modules. In fact, `x/bank` already has existing migrations. See `RegisterServices`. You'll find various `RegisterMigration` calls there. We need to do the same for this. So we have to remove `genState.MigrateSendEnabled()` :)

dwedul-figure replied:
I've already added/registered a new migration for the bank module to migrate the `Params.SendEnabled` entries into the bank module's state store: 

[x/bank/keeper/migrations.go](https://github.com/cosmos/cosmos-sdk/pull/11977/files#diff-c58b556b8ace809b59cce155271115f96a755f2d04d3ac9368c5c58e5791bc1e):
```go
// Migrate3to4 migrates x/bank storage from version 3 to 4.
func (m Migrator) Migrate3to4(ctx sdk.Context) error {
	oldParams := m.keeper.GetParams(ctx)
	m.keeper.SetAllSendEnabled(ctx, oldParams.GetSendEnabled())
	m.keeper.SetParams(ctx, banktypes.NewParams(oldParams.DefaultSendEnabled))
	return nil
}
```
and [x/bank/module.go](https://github.com/cosmos/cosmos-sdk/pull/11977/files#diff-5d9b18806fe02e80ffd38cea841e692ce20ef3d8941b135477658bad46ed9937):
```go
	if err := cfg.RegisterMigration(types.ModuleName, 3, m.Migrate3to4); err != nil {
		panic(fmt.Sprintf("failed to migrate x/bank from version 3 to 4: %v", err))
	}
```

I don't see the harm in having the `genState.MigrateSendEnabled()` call in here. All it does is make the current genesis file `Params.SendEnabled` entries work like the new `SendEnabled` entries. All that function does is move things around in that `genState` struct. It puts the `SendEnabled` entries where they're expected to be so that later in `InitGenesis`, we only need to look in one place for those entries. It allows for backwards compatibility with the old structure (defining them in Params), even in the genesis files.

alexanderbez replied:
Ahh I see! Thank you.

Re `InitGenesis`, we don't want any migrations, that's the whole point of explicit migrations. We want app devs to explicitly execute migrations and not have it happen magically via `InitGenesis`. In fact, `InitGenesis` would only in this case would only be called for a new chain, in which case there is nothing to migrate anyway.

Please remove this so we can wrap up the PR 🙏

dwedul-figure commented on x/bank/client/cli/query.go:
```diff
@@ -213,3 +214,54 @@ To query for the total supply of a specific coin denomination use:
+func GetCmdQuerySendEnabled() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "send-enabled [denom1 ...]",
+		Short: "Query for send enabled entries",
```
It's possible to get all entries by not providing any denoms. So zero or more args is valid.

dwedul-figure commented on x/bank/client/cli/query.go:
```diff
@@ -213,3 +214,54 @@ To query for the total supply of a specific coin denomination use:
+			if err != nil {
+				return err
+			}
+			return clientCtx.PrintProto(res)
```
Done.

robert-zaremba commented on proto/cosmos/bank/v1beta1/genesis.proto:
```diff
@@ -21,8 +21,11 @@ message GenesisState {
+  // denom_metadata defines the metadata of the different coins.
   repeated Metadata denom_metadata = 4 [(gogoproto.nullable) = false];
+
+  // send_enabled defines the denoms where send is enabled or disabled.
```
Or something like `// TOOD: Since: XXX`. We need to put something and be consistent with that. Otherwise we have more troubles to identify new things.

dwedul-figure replied:
Done.

I went with `// Since: cosmos-sdk 0.47` and also changed the deprecation comment to reference `cosmos-sdk 0.47`. So they're consistent and hopefully easier to find if they need to be updated due to not actually going into v0.47.

robert-zaremba reviewed (approved):
- [ ] Need to bump `bank` ConsensuVersion

pre-approving, see comments and note above

dwedul-figure commented:
Note: I don't have permissions to interact with the merge (or "squash and merge") button. So someone else will need to either enable auto-squash-and-merge, or else hit the merge button.

alexanderbez commented on proto/cosmos/bank/v1beta1/bank.proto:
```diff
@@ -10,8 +10,13 @@ option go_package = "github.com/cosmos/cosmos-sdk/x/bank/types";
+  // For genesis, use the newly added send_enabled field in the genesis object.
+  // Storage, lookup, and manipulation of this information is now in the keeper.
+  //
+  // As of cosmos-sdk 0.47, this only exists for backwards compatibility of genesis files.
```
```suggestion
  // As of cosmos-sdk 0.46, this only exists for backwards compatibility of genesis files.
```

alexanderbez commented on proto/cosmos/bank/v1beta1/genesis.proto:
```diff
@@ -21,8 +21,13 @@ message GenesisState {
+
+  // send_enabled defines the denoms where send is enabled or disabled.
+  //
+  // Since: cosmos-sdk 0.47
```
We don't know what the next version will be, in fact, we're striving towards 1.0.

dwedul-figure replied:
According to [ADR 044: Guidelines for Updating Protobuf Definitions](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-044-protobuf-updates-guidelines.md#2-non-msg-related-protobuf-definitions-may-have-new-fields-but-must-add-a-since-comment), this needs to be `0.47`.

robert-zaremba replied:
The point is to be specific and use something consistently:
* we are not 100% sure what will be the next release. We have already many unplanned point releases (0.41, 0.43, 0.45 ...).
* so let's use whatever makes sense , it can be 1.0, 0.47, XXX  - something that we can easily search and update when the next release will be clear. And be consistent. Searching for 0.46 will be confusing. Searching for 0.47 or v1.0 in proto is more precise I think.

alexanderbez commented on x/bank/keeper/genesis.go:
```diff
@@ -10,8 +10,13 @@ import (
 
 // InitGenesis initializes the bank module's state from a given genesis state.
 func (k BaseKeeper) InitGenesis(ctx sdk.Context, genState *types.GenesisState) {
+
```
```suggestion
```

alexanderbez commented on x/bank/keeper/genesis.go:
```diff
@@ -48,10 +53,13 @@ func (k BaseKeeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
 		k.GetAllDenomMetaData(ctx),
+		k.GetAllSendEnabledEntries(ctx),
 	)
+	rv.MigrateSendEnabled()
```
```suggestion
```

dwedul-figure replied:
This is needed so that, if a genesis file is created and there's somehow still `Params.SendEnabled` entries, they're moved to the correct place.

alexanderbez reviewed (commented):
Looks good! I just don't know if we should be putting "0.47" everywhere since we don't know what the next version will be.

alexanderbez commented:
Hey @dwedul-figure, thanks for pushing this over the finish line! However, I think this PR was merged just a bit prematurely. The core changes in the PR look great, but there's still a few things that need to be polished up prior to being merged into `main` IMO.

In general, I'd like to see an approval from at least one core SDK team member prior to merging substantial units of work. Would you be open to addressing the last few remaining items?

dwedul-figure commented:
> Hey @dwedul-figure, thanks for pushing this over the finish line! However, I think this PR was merged just a bit prematurely. The core changes in the PR look great, but there's still a few things that need to be polished up prior to being merged into `main` IMO.
> 
> In general, I'd like to see an approval from at least one core SDK team member prior to merging substantial units of work. Would you be open to addressing the last few remaining items?

I guess I didn't realize that @robert-zaremba wasn't part of the core sdk team.

The review I dismissed as "All concerns have been addressed." was about the version string to use in comments. But that discussion looked like it was over, settled on using `0.47`.

What other changes do you feel were needed?

robert-zaremba commented:
The proto comments in this PR follow the ADR 44.

alexanderbez commented:
There's still usage of default params which should not exist, also there is some general structure cleanup we should perform.

dwedul-figure commented:
> There's still usage of default params which should not exist, also there is some general structure cleanup we should perform.

I'm asking for specifics here.

At this point, I feel the best way forward is for you to create an issue and I'll open a new PR just for those changes.

And I'm feeling a bit defensive here, so I'd like to point out that neither of those concerns were brought up in the four weeks that this PR was open. Continually "And Then"ing a PR is disrespectful of everyone's time.

alexanderbez commented on x/bank/keeper/migrations.go:
```diff
@@ -25,3 +26,11 @@ func (m Migrator) Migrate1to2(ctx sdk.Context) error {
+func (m Migrator) Migrate3to4(ctx sdk.Context) error {
+	oldParams := m.keeper.GetParams(ctx)
+	m.keeper.SetAllSendEnabled(ctx, oldParams.GetSendEnabled())
+	m.keeper.SetParams(ctx, banktypes.NewParams(oldParams.DefaultSendEnabled))
```
Why are we calling this line?

alexanderbez replied:
It is deprecated -- that's the point of this feature. Deprecate the `DefaultSendEnabled` and use a more strongly structured list. In fact, the Proto schema states it's deprecated.

alexanderbez commented on x/bank/keeper/send.go:
```diff
@@ -67,12 +75,16 @@ func (k BaseSendKeeper) GetParams(ctx sdk.Context) (params types.Params) {
+	if len(params.SendEnabled) > 0 {
+		k.SetAllSendEnabled(ctx, params.SendEnabled)
+	}
+	p := types.NewParams(params.DefaultSendEnabled)
```
Why are we modifying default/legacy params?

dwedul-figure replied:
The `Params.DefaultSendEnabled` field is not deprecated.

alexanderbez commented on x/bank/keeper/send.go:
```diff
@@ -305,17 +317,30 @@ func (k BaseSendKeeper) setBalance(ctx sdk.Context, addr sdk.AccAddress, balance
+	}
+	store := ctx.KVStore(k.storeKey)
+	haveDefault := false
+	var defaultVal bool
```
This is an incorrect implementation. Again, we should not be referencing or using the legacy params at all.

dwedul-figure replied:
Here again, the `Params.DefaultSendEnabled` value is not depreciated and is still to be used. It is the fallback for cases when there is no `SendEnabled` entry for a denom. That functionality hasn't changed. The extra code in here just makes it so that the bank module's `Params` only need to be looked up once during this function call (rather than once per coin).

alexanderbez commented on x/bank/keeper/send.go:
```diff
@@ -328,3 +353,113 @@ func (k BaseSendKeeper) BlockedAddr(addr sdk.AccAddress) bool {
+
+// IsSendEnabledDenom returns the current SendEnabled status of the provided denom.
+func (k BaseSendKeeper) IsSendEnabledDenom(ctx sdk.Context, denom string) bool {
+	return k.getSendEnabledOrDefault(ctx.KVStore(k.storeKey), denom, func() bool { return k.GetParams(ctx).DefaultSendEnabled })
```
This needs to be undone -- again, legacy or default params should not be used.

dwedul-figure replied:
The `Params.DefaultSendEnabled` field is not deprecated. Only the `Params.SendEnabled` field.

There still needs to be a default defined, and currently, all chains have it defined in the bank module's `Params`. And it works just fine being there, so there's no reason to change that.

alexanderbez reviewed (commented):
I commented on a few points, but the overall theme is that the legacy params should not be referenced or used at all except for the migration.

alexanderbez commented:
I understand @dwedul-figure, but keep in mind our team is small and we have tons of going on. A PR like this is relatively substantial and it takes time to thoroughly review.

I've add some additional comments.

marbar3778 commented:
This may fall on me as well for not including it in the teams sprint. the past two weeks were focused on the sprint board, which can be followed here: https://github.com/orgs/cosmos/projects/26/views/22. If your pr is sitting around for a while, please feel free to ping a member of the team to try to get your review on the sprint board. This will lead to faster responses and reviews

alexanderbez commented:
We will keep this PR and have follow up PRs with minor tweaks and adjustments. Thanks for the valuable contribution @dwedul-figure :)
